ext/dom/tests/modern/common/template_participation.phpt

```diff
@@ -0,0 +1,83 @@
+--TEST--
+<template> element contents do not participate in DOM
+--EXTENSIONS--
+dom
+--FILE--
+<?php
+
+$html = <<<HTML
+<!DOCTYPE html>
+<html>
+    <body>
+        <template>a<div>foo</div>b</template>
+    </body>
+</html>
+HTML;
+$dom = Dom\HTMLDocument::createFromString($html);
+$template = $dom->body->firstElementChild;
+
+echo "=== Manipulation ===\n";
+
+echo "First child of template: ";
+var_dump($template->firstChild?->nodeName);
+$template->append($dom->createElement('invisible'));
+
+echo "First child of template after appending: ";
+var_dump($template->firstChild->nodeName);
+$template->innerHTML = $template->innerHTML;
+echo "Inner HTML after idempotent modification: ";
+var_dump($template->innerHTML);
+echo "Selector should not find div element in shadow DOM: ";
+var_dump($template->querySelector('div'));
+
+echo "XPath should not find div element in shadow DOM:\n";
+$xpath = new Dom\XPath($dom);
+var_dump($xpath->query('//div'));
+
+echo "=== HTML serialization ===\n";
+echo $dom->saveHTML(), "\n";
+echo "=== HTML serialization of <template> ===\n";
+echo $dom->saveHTML($template), "\n";
+echo "=== XML serialization ===\n";
+echo $dom->saveXML(), "\n";
+echo "=== XML serialization of <template> ===\n";
+echo $dom->saveXML($template), "\n";
+
+// Should not crash
+$template->remove();
+unset($template);
+
+echo "=== Creating a new template should not leak the old contents ===\n";
+$template = $dom->createElement('template');
+var_dump($template->innerHTML);
+
+?>
+--EXPECT--
+=== Manipulation ===
+First child of template: NULL
```

Comment:
I'm confused here, why is the first child NULL?

Reply:
Templated content doesn't participate in the DOM tree, so `$template` doesn't actually have any children.

Reply:
Right, this is somewhat confusing, but makes sense